akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala

```diff
@@ -984,31 +997,176 @@ abstract class GraphStageLogic private[stream] (val inCount: Int, val outCount:
+   * due to stream completion.
+   *
    * This object can be cached and reused within the same [[GraphStageLogic]].
    */
   final def getAsyncCallback[T](handler: T ⇒ Unit): AsyncCallback[T] = {
-    new AsyncCallback[T] {
-      override def invoke(event: T): Unit =
-        interpreter.onAsyncInput(GraphStageLogic.this, event, handler.asInstanceOf[Any ⇒ Unit])
+    val result = new ConcurrentAsyncCallback[T](handler)
+    concurrentAsyncCallback.add(result)
+    if (_interpreter != null) result.onStart()
+    result
+  }
+
+  /**
+    * ConcurrentAsyncCallback allows to call [[invoke()]] and [[invokeWithPromise()]] with event attribute.
+    * This event will be sent to the stream and handler will be called with this attribute in thread safe manner.
+    *
+    * Class state can be changed by [[GraphStage]] and from "Real World" (other non Akka threads)
+    * * [[GraphStage]] calls [[onStart()]] - when materialization is finished and [[onStop()]] - when processing of external events stopped
+    *   because the stage is about to stop or fail.
+    * * "Real world" calls [[invoke()]] and [[invokeWithFeedback()]]. These methods have synchronization
+    *   with class state that reflects the stream state
+    *
+    * onStart sends all events that were buffered while stream was materializing.
+    * In case "Real world" added more events while initializing, onStart checks for more events in buffer when exiting and
+    * resend new events
+    *
+    * Once class is in `Initialized` state - all "Real world" calls of [[invoke()]] and [[invokeWithFeedback()]] are running
+    * as is - without any synchronization.
+    *
+    * [[GraphStage]] is called [[onStop()]] when stream is wrapping down. onStop fails all futures for events that have not yet processed
+    * [[onStop()]] puts class in `Completed` state
+    * "Real world" calls of [[invokeWithFeedback()]] always return failed promises for `Completed` state
+    */
+  private class ConcurrentAsyncCallback[T](handler: T ⇒ Unit) extends AsyncCallback[T] {
+
+    sealed trait State
+    //waiting for materialization completion
+    private case class Pending(pendingEvents: List[Event]) extends State
+    //GraphStage sending all events to stream
+    private case object Initializing extends State
+    //stream is initialized and so no threads can just send events without any synchronization overhead
+    private case object Initialized extends State
+    //stage has been shut down, either regularly or it failed
+    private case object Completed extends State
+    //Event with feedback promise
+    private case class Event(e: T, handlingPromise: OptionVal[Promise[Done]])
+
+    val concurrentSet = ConcurrentHashMap.newKeySet[Promise[_]]()
+
+    private[this] val currentState = new AtomicReference[State](Pending(List.empty[Event]))
+
+    //is called from stream - thread pool is managed by Akka
```

Comment:
Isn't this also potentially called from the materialization, because of the `if (_interpreter != null) result.start()` part in `getAsyncCallback`?

Reply:
per my understanding materialization is happening in akka thread.
I'm going to change comment to `is called from the owning [[GraphStage]]`